![title](laptop-price.jpg)

# Laptop Prices

## Project Overview

This dataset was adapted from the Laptop Prices dataset on Kaggle. We changed the IDs and made the prices integers.

Here is a brief description of the rows:
- ID: A unique identifier for the laptop.
- Company: The name of the company that produces the laptop.
- Product: The name of the laptop.
- TypeName: The type of laptop.
- Inches: The size of the screen in inches.
- ScreenResolution: The resolution of the screen.
- CPU: The laptop CPU.
- RAM: The amount of RAM in the laptop.
- Memory: The size of the hard drive.
- GPU: The graphics card name.
- OpSys: The name of the operating system.
- Weight: The laptop weight.
- Price: The price of the laptop.

## Loading Data

In [1]:
import pandas as pd
import numpy as np
import csv
import matplotlib.pyplot as plt

In [2]:
file = open('laptops.csv', encoding='utf8')
read_file = csv.reader(file)
laptop = list(read_file)

In [3]:
header = laptop[0]
print(header)

['Id', 'Company', 'Product', 'TypeName', 'Inches', 'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight', 'Price']


In [4]:
rows = laptop[1:]
print(rows[0:5])

[['6571244', 'Apple', 'MacBook Pro', 'Ultrabook', '13.3', 'IPS Panel Retina Display 2560x1600', 'Intel Core i5 2.3GHz', '8GB', '128GB SSD', 'Intel Iris Plus Graphics 640', 'macOS', '1.37kg', '1339'], ['7287764', 'Apple', 'Macbook Air', 'Ultrabook', '13.3', '1440x900', 'Intel Core i5 1.8GHz', '8GB', '128GB Flash Storage', 'Intel HD Graphics 6000', 'macOS', '1.34kg', '898'], ['3362737', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', '575'], ['9722156', 'Apple', 'MacBook Pro', 'Ultrabook', '15.4', 'IPS Panel Retina Display 2880x1800', 'Intel Core i7 2.7GHz', '16GB', '512GB SSD', 'AMD Radeon Pro 455', 'macOS', '1.83kg', '2537'], ['8550527', 'Apple', 'MacBook Pro', 'Ultrabook', '13.3', 'IPS Panel Retina Display 2560x1600', 'Intel Core i5 3.1GHz', '8GB', '256GB SSD', 'Intel Iris Plus Graphics 650', 'macOS', '1.37kg', '1803']]


## Building the database

In order to quickly access the dataset, we can build a inventory class to store the data and build additional functions to allow the user to view the information fast.

In [5]:
class Inventory():
    
    def __init__(self, csv_filename):
        with open(csv_filename, encoding='utf8') as f:
            reader = csv.reader(f)
            file = list(reader)
        
        self.header = file[0]
        self.rows = file[1:]
        for row in self.rows:
            row[-1] = int(row[-1])

In [6]:
laptop_data = Inventory('laptops.csv')

In [7]:
print(laptop_data.header)

['Id', 'Company', 'Product', 'TypeName', 'Inches', 'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight', 'Price']


In [8]:
print('Number of rows: ', len(laptop_data.rows))

Number of rows:  1303


This dataset has 1303 rows. For a real dataset, it might have millions of rows which this class approach isnecessary.

### Check Laptop ID

In [9]:
class Inventory():
    
    def __init__(self, csv_filename):
        with open(csv_filename, encoding='utf8') as f:
            reader = csv.reader(f)
            file = list(reader)
        
        self.header = file[0]
        self.rows = file[1:]
        for row in self.rows:
            row[-1] = int(row[-1])
            
    def get_laptop_from_id(self, laptop_id):
        for row in self.rows:
            if row[0] == laptop_id:
                return row
        return None

In [10]:
laptop_data = Inventory('laptops.csv')

In [11]:
print(laptop_data.get_laptop_from_id('3362737'))

['3362737', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', 575]


In [12]:
print(laptop_data.get_laptop_from_id('3362736'))

None


Though this approach is working, but it takes O(N) time complexity to finish this simple task. One way to imporve the time efficiency is to build a dictionary which the key is the laptop id and the value is the row.

### Optimize the Checking Method

In [13]:
class Inventory():
    
    def __init__(self, csv_filename):
        with open(csv_filename, encoding='utf8') as f:
            reader = csv.reader(f)
            file = list(reader)
        
        self.header = file[0]
        self.rows = file[1:]
        for row in self.rows:
            row[-1] = int(row[-1])
        
        # Add id dictionary
        self.id_to_row = {}
        for row in self.rows:
            self.id_to_row[row[0]] = row
        
    def get_laptop_from_id(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None

In [14]:
laptop_data = Inventory('laptops.csv')

In [15]:
print(laptop_data.get_laptop_from_id('3362737'))

['3362737', 'HP', '250 G6', 'Notebook', '15.6', 'Full HD 1920x1080', 'Intel Core i5 7200U 2.5GHz', '8GB', '256GB SSD', 'Intel HD Graphics 620', 'No OS', '1.86kg', 575]


In [16]:
print(laptop_data.get_laptop_from_id('3362736'))

None


### Compare Two ID Check Methods

Modified the functions a little in order to allow us to compare both methods in side the same class.

In [17]:
class Inventory():
    
    def __init__(self, csv_filename):
        with open(csv_filename, encoding='utf8') as f:
            reader = csv.reader(f)
            file = list(reader)
        
        self.header = file[0]
        self.rows = file[1:]
        for row in self.rows:
            row[-1] = int(row[-1])
        
        # Add id dictionary
        self.id_to_row = {}
        for row in self.rows:
            self.id_to_row[row[0]] = row
            
    def get_laptop_from_id(self, laptop_id):
        for row in self.rows:
            if row[0] == laptop_id:
                return row
        return None
    
    def get_laptop_from_id_fast(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None

In [18]:
import time
import random

Generate a list with 10,000 random values between "1,000,000" and "9,999,999" (ID range) to test the runtime of each method.

In [19]:
ids = [str(random.randint(1000000, 9999999)) for _ in range(10000)]

In [20]:
laptop_data = Inventory('laptops.csv')

In [21]:
total_time_no_dict = 0
total_time_dict = 0

In [22]:
for id in ids:
    start = time.time()
    laptop_data.get_laptop_from_id(id)
    end = time.time()
    total_time_no_dict += (end-start)

In [23]:
for id in ids:
    start = time.time()
    laptop_data.get_laptop_from_id_fast(id)
    end = time.time()
    total_time_dict += (end-start)

In [24]:
print('Total Time without Dictionary: ', total_time_no_dict)
print('Total Time with Dictionary: ', total_time_dict)

Total Time without Dictionary:  0.6001698970794678
Total Time with Dictionary:  0.0019998550415039062


We can find that without using a dictionary method, the program will spend about 100 times more time to complete the task.

## Gift Card Problem

ometimes, your store offers a promotion where you give a gift card. A customer can use the gift to buy up to two laptops. To avoid having to keep track of what was already spent, the gift card has a single time usage. This means that, even if there is leftover money, it cannot be used anymore.
The prices of these laptops are 1339, 898, and 575. Say we offered a gift card of 2500. Since a customer can buy, at most, two laptops with a gift card, the maximum they can spend is 2237 (1339 plus 898). Therefore, they might feel cheated because, no matter how they spend their gift card, they cannot spend the full 2500.

You don't want to make a customer feel cheated, so whenever you issue a gift card, you want to make sure that there is at least one way to spend it in full. In other words, before issuing a gift card for D dollars, you want to make sure that either there is a laptop that costs exactly D dollars or two laptops whose costs add up to precisely D dollars.

Write a function that, given a dollar amount, checks whether it is possible to spend precisely that amount by purchasing up to two laptops.

In [25]:
class Inventory():
    
    def __init__(self, csv_filename):
        with open(csv_filename, encoding='utf8') as f:
            reader = csv.reader(f)
            file = list(reader)
        
        self.header = file[0]
        self.rows = file[1:]
        for row in self.rows:
            row[-1] = int(row[-1])
        
        self.id_to_row = {}
        for row in self.rows:
            self.id_to_row[row[0]] = row
        
    def get_laptop_from_id(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    
    def check_promotion_dollars(self, dollars):
        for row in self.rows:
            if row[-1] == dollars:
                return True
        for row1 in self.rows:
            for row2 in self.rows:
                if row1[-1] + row2[-1] == dollars:
                    return True
        return False  

In [26]:
laptop_data = Inventory('laptops.csv')

In [27]:
print(laptop_data.check_promotion_dollars(1000))

True


In [28]:
print(laptop_data.check_promotion_dollars(442))

False


### Optimize the Solution

Since we only care about whether or not there is a solution, we can store all laptops prices in a set when we initialize the inventory. Then we can check in constant time whether there is a laptop with a given price.

In [29]:
class Inventory():
    
    def __init__(self, csv_filename):
        with open(csv_filename, encoding='utf8') as f:
            reader = csv.reader(f)
            file = list(reader)
        
        self.header = file[0]
        self.rows = file[1:]
        for row in self.rows:
            row[-1] = int(row[-1])
        
        # Add price set
        self.prices = set()
        for row in self.rows:
            self.prices.add(row[-1])
        
        # Add id dictionary
        self.id_to_row = {}
        for row in self.rows:
            self.id_to_row[row[0]] = row
        
    def get_laptop_from_id(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    
    def check_promotion_dollars(self, dollars):
        if dollars in self.prices:
            return True
            
        for row in self.rows:
            if (dollars - row[-1]) in self.prices:
                return True
        return False

In [30]:
laptop_data = Inventory('laptops.csv')

In [31]:
print(laptop_data.check_promotion_dollars(1000))

True


In [32]:
print(laptop_data.check_promotion_dollars(442))

False


### Compare Two Gift Card Methods

Modified the functions a little in order to allow us to compare both methods in side the same class.

In [33]:
class Inventory():
    
    def __init__(self, csv_filename):
        with open(csv_filename, encoding='utf8') as f:
            reader = csv.reader(f)
            file = list(reader)
        
        self.header = file[0]
        self.rows = file[1:]
        for row in self.rows:
            row[-1] = int(row[-1])
        
        # Add price set
        self.prices = set()
        for row in self.rows:
            self.prices.add(row[-1])
        
        # Add id dictionary
        self.id_to_row = {}
        for row in self.rows:
            self.id_to_row[row[0]] = row
        
    def get_laptop_from_id(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    
    def check_promotion_dollars_fast(self, dollars):
        if dollars in self.prices:
            return True
            
        for row in self.rows:
            if (dollars - row[-1]) in self.prices:
                return True
        return False
    
    def check_promotion_dollars(self, dollars):
        for row in self.rows:
            if row[-1] == dollars:
                return True
        for row1 in self.rows:
            for row2 in self.rows:
                if row1[-1] + row2[-1] == dollars:
                    return True
        return False  

In [34]:
laptop_data = Inventory('laptops.csv')

In [35]:
prices = [random.randint(100, 5000) for _ in range(100)]

In [36]:
total_time_no_set = 0
total_time_set = 0

In [37]:
for price in prices:
    start = time.time()
    laptop_data.check_promotion_dollars(price)
    end = time.time()
    total_time_no_set += (end - start)

In [38]:
for price in prices:
    start = time.time()
    laptop_data.check_promotion_dollars_fast(price)
    end = time.time()
    total_time_set += (end - start)

In [39]:
print('Total Time without Set: ', total_time_no_set)
print('Total Time with Set: ', total_time_set)

Total Time without Set:  0.8505070209503174
Total Time with Set:  0.001001596450805664


## Find All Laptops under a Certain Price

In [40]:
# Sort the row by price
def row_price(row):
    return row[-1]

class Inventory():
    
    def __init__(self, csv_filename):
        with open(csv_filename, encoding='utf8') as f:
            reader = csv.reader(f)
            file = list(reader)
        
        self.header = file[0]
        self.rows = file[1:]
        for row in self.rows:
            row[-1] = int(row[-1])
        
        # Add price set
        self.prices = set()
        for row in self.rows:
            self.prices.add(row[-1])
        
        # Add id dictionary
        self.id_to_row = {}
        for row in self.rows:
            self.id_to_row[row[0]] = row
            
        self.rows_by_price = sorted(self.rows, key=row_price)
    
    # Get laptop ID
    def get_laptop_from_id(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    
    # Check promotion price
    def check_promotion_dollars(self, dollars):
        if dollars in self.prices:
            return True
            
        for row in self.rows:
            if (dollars - row[-1]) in self.prices:
                return True
        return False
    
    # Find laptops under a budget
    def find_first_laptop_more_expensive(self, target_price):
        range_start = 0
        range_end = len(self.rows_by_price) - 1
        
        while range_start < range_end:
            range_mid = (range_start + range_end) // 2
            price = self.rows_by_price[range_mid][-1]
            if target_price >= price:
                range_start = range_mid + 1
            else:
                range_end = range_mid
        if self.rows_by_price[range_start][-1] <= target_price:
            return -1
        return range_start

In [41]:
laptop_data = Inventory('laptops.csv')

In [42]:
print(laptop_data.find_first_laptop_more_expensive(1000))

683


In [43]:
print(laptop_data.find_first_laptop_more_expensive(10000))

-1


## Taptops within Budget

In [44]:
# Sort the row by price
def row_price(row):
    return row[-1]

class Inventory():
    
    def __init__(self, csv_filename):
        with open(csv_filename, encoding='utf8') as f:
            reader = csv.reader(f)
            file = list(reader)
        
        self.header = file[0]
        self.rows = file[1:]
        for row in self.rows:
            row[-1] = int(row[-1])
        
        # Add price set
        self.prices = set()
        for row in self.rows:
            self.prices.add(row[-1])
        
        # Add id dictionary
        self.id_to_row = {}
        for row in self.rows:
            self.id_to_row[row[0]] = row
            
        self.rows_by_price = sorted(self.rows, key=row_price)
    
    # Get laptop ID
    def get_laptop_from_id(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    
    # Check promotion price
    def check_promotion_dollars(self, dollars):
        if dollars in self.prices:
            return True
            
        for row in self.rows:
            if (dollars - row[-1]) in self.prices:
                return True
        return False
    
    # Find laptops under a budget
    def find_first_laptop_more_expensive(self, target_price):
        range_start = 0
        range_end = len(self.rows_by_price) - 1
        
        while range_start < range_end:
            range_mid = (range_start + range_end) // 2
            price = self.rows_by_price[range_mid][-1]
            if target_price >= price:
                range_start = range_mid + 1
            else:
                range_end = range_mid
        if self.rows_by_price[range_start][-1] <= target_price:
            return -1
        return range_start
    
    def maximum_price(self, target_price):
        expensive_laptop_index = self.find_first_laptop_more_expensive(target_price)
        if expensive_laptop_index == -1:
            return self.rows_by_price
        elif expensive_laptop_index == 0:
            return "No match can be found"
        else:
            # Including that target price
            all_laptops = self.rows_by_price[0:expensive_laptop_index]
            return all_laptops

In [45]:
laptop_data = Inventory('laptops.csv')

In [46]:
print(laptop_data.maximum_price(2))

No match can be found


In [47]:
print(laptop_data.maximum_price(1000))

[['3564228', 'Acer', 'C740-C9QX (3205U/2GB/32GB/Chrome', 'Netbook', '11.6', '1366x768', 'Intel Celeron Dual Core 3205U 1.5GHz', '2GB', '32GB SSD', 'Intel HD Graphics', 'Chrome OS', '1.3kg', 174], ['7667029', 'Asus', 'Vivobook E200HA', 'Netbook', '11.6', '1366x768', 'Intel Atom x5-Z8350 1.44GHz', '2GB', '32GB Flash Storage', 'Intel HD Graphics 400', 'Windows 10', '0.98kg', 191], ['1478754', 'Vero', 'V131 (X5-Z8350/4GB/32GB/FHD/W10)', 'Notebook', '13.3', 'Full HD 1920x1080', 'Intel Atom X5-Z8350 1.44GHz', '4GB', '32GB Flash Storage', 'Intel HD Graphics 400', 'Windows 10', '1.35kg', 196], ['4366200', 'Asus', 'E402WA-GA010T (E2-6110/2GB/32GB/W10)', 'Notebook', '14', '1366x768', 'AMD E-Series E2-6110 1.5GHz', '2GB', '32GB Flash Storage', 'AMD Radeon R2', 'Windows 10', '1.65kg', 199], ['3840240', 'Acer', 'Chromebook C910-C2ST', 'Notebook', '15.6', '1366x768', 'Intel Celeron Dual Core 3205U 1.5GHz', '2GB', '16GB SSD', 'Intel HD Graphics', 'Chrome OS', '2.19kg', 199], ['7260172', 'Vero', 'K146

In [48]:
print(laptop_data.maximum_price(10000))

[['3564228', 'Acer', 'C740-C9QX (3205U/2GB/32GB/Chrome', 'Netbook', '11.6', '1366x768', 'Intel Celeron Dual Core 3205U 1.5GHz', '2GB', '32GB SSD', 'Intel HD Graphics', 'Chrome OS', '1.3kg', 174], ['7667029', 'Asus', 'Vivobook E200HA', 'Netbook', '11.6', '1366x768', 'Intel Atom x5-Z8350 1.44GHz', '2GB', '32GB Flash Storage', 'Intel HD Graphics 400', 'Windows 10', '0.98kg', 191], ['1478754', 'Vero', 'V131 (X5-Z8350/4GB/32GB/FHD/W10)', 'Notebook', '13.3', 'Full HD 1920x1080', 'Intel Atom X5-Z8350 1.44GHz', '4GB', '32GB Flash Storage', 'Intel HD Graphics 400', 'Windows 10', '1.35kg', 196], ['4366200', 'Asus', 'E402WA-GA010T (E2-6110/2GB/32GB/W10)', 'Notebook', '14', '1366x768', 'AMD E-Series E2-6110 1.5GHz', '2GB', '32GB Flash Storage', 'AMD Radeon R2', 'Windows 10', '1.65kg', 199], ['3840240', 'Acer', 'Chromebook C910-C2ST', 'Notebook', '15.6', '1366x768', 'Intel Celeron Dual Core 3205U 1.5GHz', '2GB', '16GB SSD', 'Intel HD Graphics', 'Chrome OS', '2.19kg', 199], ['7260172', 'Vero', 'K146

## Final Optimized Class

In [49]:
# Sort the row by price
def row_price(row):
    return row[-1]

class Inventory():
    
    def __init__(self, csv_filename):
        with open(csv_filename, encoding='utf8') as f:
            reader = csv.reader(f)
            file = list(reader)
        
        self.header = file[0]
        self.rows = file[1:]
        for row in self.rows:
            row[-1] = int(row[-1])
        
        # Add price set
        self.prices = set()
        for row in self.rows:
            self.prices.add(row[-1])
        
        # Add id dictionary
        self.id_to_row = {}
        for row in self.rows:
            self.id_to_row[row[0]] = row
            
        self.rows_by_price = sorted(self.rows, key=row_price)
    
    # Get laptop ID
    def get_laptop_from_id(self, laptop_id):
        if laptop_id in self.id_to_row:
            return self.id_to_row[laptop_id]
        return None
    
    # Check promotion price
    def check_promotion_dollars(self, dollars):
        if dollars in self.prices:
            return True
            
        for row in self.rows:
            if (dollars - row[-1]) in self.prices:
                return True
        return False
    
    # Find laptops under a budget
    def find_first_laptop_more_expensive(self, target_price):
        range_start = 0
        range_end = len(self.rows_by_price) - 1
        
        while range_start < range_end:
            range_mid = (range_start + range_end) // 2
            price = self.rows_by_price[range_mid][-1]
            if target_price >= price:
                range_start = range_mid + 1
            else:
                range_end = range_mid
        if self.rows_by_price[range_start][-1] <= target_price:
            return -1
        return range_start
    
    def maximum_price(self, target_price):
        expensive_laptop_index = self.find_first_laptop_more_expensive(target_price)
        if expensive_laptop_index == -1:
            return self.rows_by_price
        elif expensive_laptop_index == 0:
            return "No match can be found"
        else:
            # Including that target price
            all_laptops = self.rows_by_price[0:expensive_laptop_index]
            return all_laptops